### DISCUSSION TITLE
Terminal Suggestions

### DISCUSSION THREAD

**Discussion by advice2020** [doublecmd/doublecmd#1280] [Suggestions]
`NAMING SCHEME`
I would use “cm_Terminal....” prefix for naming new commands mentioned below, but “cm_ToggleFullscreenConsole” already exists so I went with this naming scheme.  Renaming might cause issues but maybe could rename this command to “cm_TerminalToggleFullscreen” because the Options - Layout says “Show terminal window”.  If “Console” is the correct term, then maybe “cm_ConsoleToggleFullscreen”

`SUGGESTIONS`
1.) Add new DC command to toggle show/hide terminal
Name: “cm_ToggleShowConsole”
(the terminal is not something that is necessariliy needed all the time)
Currently can only do this by having to go into Options - Layout and check/uncheck box.

2.) Add new DC command to be able to toggle current view (height) / hide terminal
Name: “cm_ToggleViewConsole”   or   “cm_ToggleCurrentViewConsole”
This would be similar to the DC command “cm_ToggleFullscreenConsole” which contains two modes: full screen & hidden.
If Option is set to “show terminal” the “Ctrl+O” is the default hotkey that toggles between them.
Instead of "fullscreen" this new command would remember the height of the terminal (set by the user) and allow toggling between that height and hide, which would allow users to be able to view the file pane & the terminal at the same time.

3.) When terminal is visible, make the horizontal height line with "....." grip appear at the top of the terminal (below the file pane status bar)
(currently remains at the bottom)

4.) Allow the horizontal height line with grip at top of terminal to be able to be adjusted....
-All the way to the top of the screen (like how “fullscreen” allows)
-All the way to the bottom of the screen (like how "hide" allows)

5.) Add a hotkey to allow focus in & out of terminal.
Name: “cm_ToggleFocusConsole”
(Might be similar to Quick Preview Pane where this is not possible?)

6.) Allow ability to highlight text items with mouse drag and be able to copy text to clipboard.
Add context menu like terminal emulators have with "copy" etc.
Allow to be able to “paste” into command line as well.
Also allow ability to drag & drop items into the terminal window.

7.) Allow when you type in "exit" the DC terminal pane closes.

`POSSIBLE BUGS / GRAPHICAL ISSUES`
A.) I think this is just how this was designed to work, but when you load up DC and the terminal is at specific height the command prompt loads up fine. If you adjust the height manually after that, the command prompt dissappears leaving just the “$” leaves blank black space above. Hitting "Enter" key brings the command prompt back.
This also occurs if you use the “toggle full screen mode” to hide the terminal, if you drag up from bottom of screen to see it there is just a black screen with a “$”

Thank You to anyone who reads this
Hopefully some of these ideas will be helpful

**Comment by RoyiAvital** [comment]
It would be great if the path of the active tab in each side will be assigned to environment variable.
Then using the terminal it will be easy to apply inter sides operations.
Maybe even scripts which will be aware of the current state of path on each side.

**Reply by JonnyTech** [reply]
https://doublecmd.github.io/doc/en/variables.html#basicsuffix

**Reply by AngelQC** [reply]
These do not work, at least on linux, $ps shows nothing and %ps shows "%ps", as suspected.

**Comment by RoyiAvital** [comment]
Hi,
When I look on the changes for version `1.11` I was mostly looking for:

![image](https://github.com/doublecmd/doublecmd/assets/7347975/f494d0e2-2bd3-4427-b219-efb4e246584b)

Yet I don't see it in the program.
I thought the concept would be like the built in terminal in VSCode.

There is the line of command which always has been there.
Using `F9` launches the system terminal.

What am I missing?

**Reply by j2969719** [reply]
press `Ctrl+O`

**Comment by advice2020** [comment]
I updated my original post above if anyone is interested.

**Comment by Vladimyr0** [comment]
copy and paste features are absolutely must have in embedded console! 
I vote for it :)
